src/routes/fastestpath.py

```diff
@@ -20,36 +25,52 @@ def fastest_path(
     :param G: A graph corresponding to the road network
     :return: A list of routes where each route corresponds to the origin point at the same index.
     """
-    logging.info("Routing shortest path to safety for all origin points")
+    has_path_been_calculated = dict((node, False) for node in origin_points)
     routes = []
 
+    logging.info("Routing fastest path to safety for all origin points")
+
     for origin in tqdm(origin_points):
+        if has_path_been_calculated[origin]:
+            continue  # path has already been calculated in another iteration
+        if len(list(G.neighbors(origin))) == 0:
+            logging.info(f"Node {origin} has no neighbors")
+            continue  # Skip if the origin node doesn't have neighbors
         sptSet = dict((node, False) for node in list(G.nodes))
         dist: list[tuple[float, str]] = [(float("inf"), node) for node in G.nodes]
         node_priority = {node: float("inf") for node in G.nodes}
         hq.heapify(dist)
         predecessor: dict[str, str | None] = {
             node: None for node in G.nodes
-        }  # To track shortest path
+        }  # To track fastest path
 
         update_priority(dist, node_priority, origin, 0)
-
         while dist:
-            priority, smallest_node = hq.heappop(dist)
-
+            priority, smallest_node = hq.heappop(
+                dist
+            )  # popping the node with the current fastest path to origin
+            if priority == float("inf"):
+                logging.info(
+                    f"Node {origin} cannot reach any nodes outside the dangerzone"
+                )
+                break
             if (
                 priority == node_priority[smallest_node]
             ):  # Only use values that are not outdated
                 if not sptSet[smallest_node]:
                     sptSet[smallest_node] = True
                     for _, neighbour, edge_data in G.edges(
                         smallest_node, data=True
-                    ):  # Multiple edges between two nodes possible
+                    ):  # looping through all adjacent vertices
                         length = edge_data.get(
                             "length", float("inf")
-                        )  # Edge length in meters
+                        )  # Default weight to inf, weight of edge between smallest_node and neighbour
+
                         maxspeed = edge_data.get("maxspeed", 50)  # speed limit km/h
-                        maxspeed = maxspeed * 0.27778  # converting km/h to m/s
+                        maxspeed = (
+                            maxspeed[0] if isinstance(maxspeed, list) else maxspeed
+                        )  # take the first speed limit in case there are more
+                        maxspeed = float(maxspeed) * 0.27778  # converting km/h to m/s
```

Comment:
Will crash if `maxspeed` is non-numeric. Consider wrapping in a `try-except` block, logging invalid values as warnings, and setting the default to 50 km/h